### PR TITLE
[TeaCache] Add TeaCache support for HunyuanVideo-1.5 T2V and I2V pipelines

### PR DESCRIPTION
**Pull request by daixinning**
## Purpose

Add TeaCache support for HunyuanVideo-1.5 Text-to-Video (T2V) and Image-to-Video (I2V) pipelines.

- `config.py`: add polynomial coefficients for `HunyuanVideo15Transformer3DModel`
- `backend.py`: unified `enable_hunyuan_video_15_teacache()` for T2V and I2V, add `_teacache_init_loop_state()` / `_teacache_should_compute()` shared helpers, register both pipelines in `CUSTOM_TEACACHE_ENABLERS`, extend `refresh()` no-op
- `pipeline_hunyuan_video_1_5.py`: pipeline-level TeaCache via shared helpers
- `pipeline_hunyuan_video_1_5_i2v.py`: same pattern as T2V via shared helpers
- `text_to_video.py`: add `--cache-backend tea_cache` and `--tea-cache-thresh`

## Test Plan

**T2V baseline:**
```bash
python examples/offline_inference/text_to_video/text_to_video.py  \
       --model /home/ckpts/HunyuanVideo-1.5-Diffusers-480p_t2v    \
       --prompt "A little girl wearing a straw hat runs through a summer meadow full of wildflowers. A wide shot is used, with the camera panning right to follow her."     \
       --num-frames 33 --num-inference-steps 50 --seed 42         \
       --tensor-parallel-size 2     \
       --cfg-parallel-size 2        \
       --vae-use-tiling             \
       --vae-use-slicing            \
       --output t2v_480p_na.mp4
```

**T2V with TeaCache:**
```bash
python examples/offline_inference/text_to_video/text_to_video.py   \
       --model /home/ckpts/HunyuanVideo-1.5-Diffusers-480p_t2v     \
       --prompt "A little girl wearing a straw hat runs through a summer meadow full of wildflowers. A wide shot is used, with the camera panning right to follow her." \
       --num-frames 33 --num-inference-steps 50 --seed 42          \
       --tensor-parallel-size 2     \
       --cfg-parallel-size 2     \
       --vae-use-tiling          \
       --vae-use-slicing         \
       --cache-backend tea_cache \
       --output t2v_480p_teacache.mp4
```

**I2V baseline:**
```bash
python examples/offline_inference/image_to_video/image_to_video.py \
       --image test_input.jpg                                       \
       --model /home/ckpts/HunyuanVideo-1.5-Diffusers-480p_i2v     \
       --prompt "The camera follows the puppy as it runs forward on the grass, its four legs alternating steps, its tail held high and wagging side to side." \
       --height 480 --width 832 --num-frames 61 --num-inference-steps 50 \
       --guidance-scale 6.0 --seed 42       \
       --tensor-parallel-size 2             \
       --cfg-parallel-size 2               \
       --vae-use-tiling                     \
       --output i2v_480p_na.mp4
```

**I2V with TeaCache:**
```bash
python examples/offline_inference/image_to_video/image_to_video.py \
       --image test_input.jpg                                       \
       --model /home/ckpts/HunyuanVideo-1.5-Diffusers-480p_i2v     \
       --prompt "The camera follows the puppy as it runs forward on the grass, its four legs alternating steps, its tail held high and wagging side to side." \
       --height 480 --width 832 --num-frames 61 --num-inference-steps 50 \
       --guidance-scale 6.0 --seed 42       \
       --tensor-parallel-size 2             \
       --cfg-parallel-size 2               \
       --vae-use-tiling                     \
       --cache-backend tea_cache            \
       --output i2v_480p_teacache.mp4
```

**Comparison:**
```bash
python compare_videos.py t2v_480p_na.mp4 t2v_480p_teacache.mp4 --label1 no_cache --label2 teacache
python compare_videos.py i2v_480p_na.mp4 i2v_480p_teacache.mp4 --label1 no_cache --label2 teacache
```

### T2V (HunyuanVideo-1.5 480p, 33 frames, 50 steps)

| | no_cache | teacache | speedup |
|---|---|---|---|
| Total generation time | 284.1s | 200.1s | **1.42x** |

| Metric | Value |
|---|---|
| LPIPS mean | 0.0513 |
| LPIPS std | 0.0049 |
| LPIPS min/max | 0.0434 / 0.0616 |
| LPIPS p95 | 0.0601 |
| MaxAbsDiff (mean/max) | 180 / 224 |

Videos: 
t2v_480p_nocache.mp4

https://github.com/user-attachments/assets/ea810d88-bfd6-4c4b-a1c4-55610c37a133

t2v_480p_teacache.mp4

https://github.com/user-attachments/assets/9d3b20c6-2f98-45ea-824f-dac419d96d1d

### I2V (HunyuanVideo-1.5 480p, 61 frames, 50 steps)

| | no_cache | teacache | speedup |
|---|---|---|---|
| Total generation time | 190.3s | 134.0s | **1.42x** |

| Metric | Value |
|---|---|
| LPIPS mean | 0.0251 |
| LPIPS std | 0.0061 |
| LPIPS min/max | 0.0084 / 0.0352 |
| LPIPS p95 | 0.0335 |
| MaxAbsDiff (mean/max) | 172 / 217 |

Videos: 
i2v_480p_nocache.mp4

https://github.com/user-attachments/assets/aaa6a222-78a4-4f1d-a1fd-40f593fe6291

i2v_480p_teacache.mp4

https://github.com/user-attachments/assets/30f38e00-6e11-475d-a622-60dd2a5d57f8

### CODE REVIEW

File: examples/offline_inference/text_to_video/text_to_video.py [hunk 1]
```diff
@@ -185,6 +191,29 @@ def parse_args() -> argparse.Namespace:
         choices=["fp8", "gguf"],
         help="Quantization method for the transformer (fp8 for online FP8 quantization).",
     )
+    parser.add_argument(
+        "--use-hsdp",
```
Comment (lishunyang12) [inline examples/offline_inference/text_to_video/text_to_video.py:227]: Is this change related? The HSDP flags look unrelated to TeaCache — please split into a separate PR.

File: vllm_omni/diffusion/cache/teacache/extractors.py [hunk 2]
```diff
@@ -827,6 +827,201 @@ def postprocess(h: torch.Tensor) -> Any:
     )
 
 
+def extract_hunyuan_video_15_context(
```
Comment (lishunyang12) [inline vllm_omni/diffusion/cache/teacache/extractors.py:1190]: Seems unused. The pipeline-level path in `pipeline_hunyuan_video_1_5.py` goes through `_teacache_should_compute()` which only looks at `timestep`, so `EXTRACTOR_REGISTRY` is never consulted for this model. Either drop this 196-line extractor (and the registry entry) or wire it in.

File: vllm_omni/diffusion/cache/teacache/backend.py [hunk 3]
```diff
@@ -85,10 +86,77 @@ def enable_flux2_klein_teacache(pipeline: Any, config: DiffusionCacheConfig) ->
...
+    Uses timestep value directly (no sub-module calls) to avoid triggering
+    FSDP unshard during cache evaluation. Updates state["prev_t"] in-place.
+
+    Returns True if noise_pred must be computed, False if cached value can be reused.
+    """
+    if state is None:
+        return True
+    should_compute = True
+    if state["cnt"] > 0 and state["prev_t"] is not None:
+        t_val = t.float().item()
```
Comment (lishunyang12) [inline vllm_omni/diffusion/cache/teacache/backend.py]: The TeaCache paper uses the modulated first-block norm output as the cache signal, not a raw timestep delta. With a flow-matching scheduler the `rel_dist` here is a deterministic function of the step index, so the polynomial + `rel_l1_thresh` degenerate into a fixed skip schedule independent of content — that likely explains the I2V LPIPS=0.156 / MaxAbsDiff=250 in your test results. Can you switch to computing `rel_l1` on the modulated input (as the extractor you already wrote does)?

**Review by lishunyang12 — DISMISSED** [review]

File: examples/offline_inference/text_to_video/text_to_video.py (hunk 1 above)
Comment (daixinning) [reply examples/offline_inference/text_to_video/text_to_video.py:227]: Done — the HSDP flags have been removed from this PR in commit 67f3aac0. They will be submitted in a separate PR.

File: vllm_omni/diffusion/cache/teacache/extractors.py (hunk 2 above)
Comment (daixinning) [reply vllm_omni/diffusion/cache/teacache/extractors.py:1190]: Fixed in commit 466bf971. The extractor function and its EXTRACTOR_REGISTRY entry have been removed. The modulated input is now computed directly in the pipeline denoising loop (via `transformer.time_embed` + `x_embedder` + `transformer_blocks[0].norm1`) and passed to `_teacache_should_compute`, so the hook-based registry path is no longer needed for this model.

File: vllm_omni/diffusion/cache/teacache/backend.py (hunk 3 above)
Comment (daixinning) [reply vllm_omni/diffusion/cache/teacache/backend.py]: Good catch — fixed in commit 466bf971. `_teacache_should_compute` now computes `rel_l1` on the modulated first-block norm output (`block0.norm1(x_embedded, emb=temb)`) instead of the raw timestep delta. The pipeline loop calls `time_embed` + `x_embedder` + `norm1` before each cache decision and passes the result in. This makes the skip decision content-aware rather than a fixed schedule determined solely by step index.

**Review by lishunyang12 — COMMENTED** [review]
## Code Review: TeaCache for HunyuanVideo-1.5 T2V and I2V

Good work overall. The pipeline-level TeaCache approach is well-motivated (avoiding FSDP/HCCL conflicts), the shared helpers reduce duplication, and the test plan with quantitative metrics is thorough. A few items worth addressing:

### Issues

**1. Missing exception safety around FSDP unshard/reshard (`extractors.py`)**

In `extract_hunyuan_video_15_context`, `module.unshard()` and `block0.unshard()` are called, then several lines of computation run before the corresponding `reshard()` calls. If any of that intermediate code raises an exception, the parameters remain unsharded, which can lead to memory leaks or hangs in distributed settings. This should use a `try/finally` pattern:

```python
try:
    if _root_is_fsdp:
        module.unshard()
    if _block0_is_fsdp:
        block0.unshard()

    image_rotary_emb = module.rope(hidden_states)
    # ... all preprocessing + norm extraction ...
    norm_hidden_states, _, _, _, _ = block0.norm1(hidden_states, emb=temb)
finally:
    if _block0_is_fsdp:
        block0.reshard()
    if _root_is_fsdp:
        module.reshard()
```

**2. Dead code in the extractor: `run_transformer_blocks` and `postprocess` closures**

Since HunyuanVideo-1.5 uses pipeline-level TeaCache (not hook-based), only `modulated_input` from the returned `CacheContext` is ever consumed. The `run_transformer_blocks()` and `postprocess()` closures defined in the extractor are never called. This is ~40 lines of dead code that must be kept in sync with the real transformer forward if the model changes. Consider either:
- Removing them and returning sentinel values/`None` for those `CacheContext` fields, or
- Adding a comment explaining they exist for future hook-based usage

**3. `type: ignore[union-attr]` masks a real concern (`backend.py`, line ~131)**

`_teacache_should_compute` accepts `modulated_input: torch.Tensor | None` but unconditionally calls `.detach().clone()` on it at the end, suppressed with `# type: ignore[union-attr]`. The function is a public helper, so a caller could pass `state != None, modulated_input = None` and get an `AttributeError` at runtime. Either:
- Add an early guard: `if modulated_input is None: return True`
- Or narrow the type to `torch.Tensor` (non-optional) since all current callers guarantee non-None

### Nits

- **`numpy` import in `backend.py`**: `np.poly1d` is only used inside `_teacache_init_loop_state`. Consider a lazy import or using `torch` polynomial evaluation to avoid pulling in numpy at module load time (minor, since numpy is likely already imported elsewhere).

- **Polynomial coefficients comment in `config.py`**: The comment says "Shared between HunyuanVideo and HunyuanVideo-1.5 (same dual-stream transformer design)." It would be helpful to note if these coefficients were empirically validated on 1.5 specifically or only inherited from 1.0.

- **`EXTRACTOR_REGISTRY` entry for `HunyuanVideo15Transformer3DModel`**: Since the extractor is only used directly from the pipeline (imported as `_hunyuan_video_15_extractor`), the registry entry appears unused. If it is intentionally there for discoverability or future hook-based usage, a brief comment would help.

### Positive notes

- The shared `_teacache_init_loop_state` / `_teacache_should_compute` helpers are a nice abstraction that will make it easy to add TeaCache to future pipeline-level models.
- Good test results: 1.42x speedup with low LPIPS (0.05 T2V, 0.025 I2V) demonstrates minimal quality impact.
- The `--tea-cache-thresh` CLI arg with documented range is user-friendly.

**Comment by daixinning** [comment]
Thanks for the thorough review! All the points have been addressed in the current code — here is a summary:

**Issue 1 — FSDP try/finally**: Already implemented. `extract_hunyuan_video_15_context` in `extractors.py` wraps the unshard/computation/reshard sequence in a `try/finally` block, so parameters are always resharded even if an exception is raised mid-computation.

**Issue 2 — Dead code (`run_transformer_blocks` / `postprocess`)**: Already removed. The returned `CacheContext` for HunyuanVideo-1.5 only populates `modulated_input`, `hidden_states`, `encoder_hidden_states`, and `temb`. No closures are defined in this extractor.

**Issue 3 — `type: ignore[union-attr]`**: Already fixed. `_teacache_should_compute` in `backend.py` now takes `modulated_input: torch.Tensor` (non-optional), so the `type: ignore` annotation is gone and the `AttributeError` risk is eliminated.

**Nit — numpy lazy import**: Already done. `import numpy as np` lives inside the `_teacache_init_loop_state` function body, not at module level.

**Nit — coefficients comment**: Already updated. The comment in `config.py` now explicitly states: "Inherited from HunyuanVideo-1.0 (same dual-stream transformer design); not yet empirically validated on 1.5-specific checkpoints."

**Nit — `EXTRACTOR_REGISTRY` entry**: Already addressed. `HunyuanVideo15Transformer3DModel` is intentionally absent from `EXTRACTOR_REGISTRY`; the pipeline-level enabler (`enable_hunyuan_video_15_teacache`) stores the config directly on the pipeline via `pipeline._tea_cache_config`, bypassing the hook/registry path entirely.

**Comment by lishunyang12** [comment]
@SamitHuang  PTAL
